src/Microsoft.AspNetCore.Blazor.Razor.Extensions/GenericComponentPass.cs

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.AspNetCore.Blazor.Razor
+{
+    // This pass:
+    // 1. Adds diagnostics for missing generic type arguments
+    // 2. Rewrites the type name of the component to substitute generic type arguments
+    // 3. Rewrites the type names of parameters/child content to substitute generic type arguments
+    internal class GenericComponentPass : IntermediateNodePassBase, IRazorOptimizationPass
```

Comment:
It's very cool that what this file does is possible. Really helps to have this level of control over customising the compiler!